File: android/Book-AndroidProgramming/QuizApp/app/src/main/java/kr/ac/ajou/heidi/quizapp/QuizActivity.java
package kr.ac.ajou.heidi.quizapp;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

public class QuizActivity extends AppCompatActivity {

    private Button mTrueButton;
    private Button mFalseButton;
    private ImageButton mNextButton;
    private ImageButton mBackButton;
    private TextView mQuestionView;
    private Button mCheatButton;
    private boolean mIsCheater;
    private static final String KEY_INDEX = "index";
    private static final int REQUEST_CODE_CHEAT = 0;

    private Question[] mQuestionBank = new Question[]{
            new Question(R.string.question_oceans, true),
            new Question(R.string.question_mideast, false),
            new Question(R.string.question_africa, false),
            new Question(R.string.question_americas, true),
            new Question(R.string.question_asia, true)
    };

    private int mCurrentIndex = 0;
    private List<Integer> mPreviousIndex = new ArrayList<>();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_quiz);

        mTrueButton = findViewById(R.id.true_button);
        mFalseButton = findViewById(R.id.false_button);
        mQuestionView = findViewById(R.id.question_text_view);
        mNextButton = findViewById(R.id.next_button);
        mBackButton = findViewById(R.id.back_button);
        mCheatButton = findViewById(R.id.cheat_button);

        mCheatButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                boolean answerIsTrue = mQuestionBank[mCurrentIndex].ismAnswerTrue();
                Intent intent = CheatActivity.newIntent(QuizActivity.this, answerIsTrue);
                startActivityForResult(intent, REQUEST_CODE_CHEAT);
            }
        });

        if (savedInstanceState != null) {
            mCurrentIndex = savedInstanceState.getInt(KEY_INDEX);
            mIsCheater = savedInstanceState.getBoolean("isCheater");
        }

        int question = mQuestionBank[mCurrentIndex].getmTextResId();
        mQuestionView.setText(question);

        mQuestionView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                updateQuestion();
            }
        });

        mTrueButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                checkAnswer(true);
            }
        });

        mFalseButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                checkAnswer(false);
            }
        });

        mNextButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mIsCheater = false;
                updateQuestion();
            }
        });

        mBackButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                backPreviousQuestion();
            }
        });


    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (resultCode != Activity.RESULT_OK) {
            return;
        }

        if (requestCode == REQUEST_CODE_CHEAT) {
            if (data == null) {
                return;
            }
            mIsCheater = CheatActivity.wasAnswerShown(data);
        }
    }

    private void updateQuestion() {
        mPreviousIndex.add(mCurrentIndex);
        mCurrentIndex = (mCurrentIndex + 1) % mQuestionBank.length;
        int question = mQuestionBank[mCurrentIndex].getmTextResId();
        mQuestionView.setText(question);
    }

    private void backPreviousQuestion() {
        mCurrentIndex = mPreviousIndex.get(mPreviousIndex.size() - 1);
        int question = mQuestionBank[mPreviousIndex.get(mPreviousIndex.size() - 1)].getmTextResId();
        mQuestionView.setText(question);
        mPreviousIndex.remove(mPreviousIndex.size() - 1);
    }

    private void checkAnswer(boolean userPressedTrue) {
        boolean answerIsTrue = mQuestionBank[mCurrentIndex].ismAnswerTrue();

        int messageResId = 0;

        if (mIsCheater || mQuestionBank[mCurrentIndex].ismIsCheater()) {
            mQuestionBank[mCurrentIndex].setmIsCheater(true);
            messageResId = R.string.judgment_toast;
        } else {
            if (userPressedTrue == answerIsTrue) {
                messageResId = R.string.correct_toast;
            } else {
                messageResId = R.string.incorrect_toast;
            }
        }

        Toast.makeText(this, messageResId, Toast.LENGTH_SHORT).show();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt(KEY_INDEX, mCurrentIndex);
        outState.putBoolean("isCheater", mIsCheater);
    }
}

File: android/Book-AndroidProgramming/GeoQuiz/app/src/main/java/kr/ac/ajou/heidi/geoquiz/QuizActivity.kt
package kr.ac.ajou.heidi.geoquiz

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import kotlinx.android.synthetic.main.activity_quiz.*
import org.jetbrains.anko.toast


class QuizActivity : AppCompatActivity() {

    private val questionBank = arrayOf(Question(R.string.question_africa, true),
            Question(R.string.question_americas, false),
            Question(R.string.question_asia, true),
            Question(R.string.question_mideast, false),
            Question(R.string.question_oceans, true)
    )

    private var currentIndex = 0
    private var previousIndex: MutableList<Int> = mutableListOf()


    companion object {
        const val KEY_INDEX = "index"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_quiz)

        if(savedInstanceState != null) {
            currentIndex = savedInstanceState.getInt(KEY_INDEX, 0)
        }


        Log.i("QuizActivity", "onCreate()")

        questionTextView.setOnClickListener {
            updateQuestion()
        }

        backButton.setOnClickListener {
            backPreviousQuestion()
        }

        nextButton.setOnClickListener {
            updateQuestion()
        }
        trueButton.setOnClickListener {
            checkAnswer(true)
        }

        falseButton.setOnClickListener {
            checkAnswer(false)
        }
        val question = questionBank[currentIndex].textResId
        questionTextView.text = getString(question)

    }

    private fun updateQuestion() {
        previousIndex.add(currentIndex)
        currentIndex = (currentIndex + 1) % questionBank.size
        val question = questionBank[currentIndex].textResId
        questionTextView.text = getString(question)
    }

    private fun backPreviousQuestion() {
        currentIndex = previousIndex.last()
        val question = questionBank[previousIndex.last()].textResId
        questionTextView.text = getString(question)
        previousIndex.removeAt(previousIndex.last())
    }

    private fun checkAnswer(userPressedTrue: Boolean) {
        val answerIsTrue = questionBank[currentIndex].answerTrue

        if (userPressedTrue == answerIsTrue) {
            toast(R.string.correct_toast)
        } else {
            toast(R.string.incorrect_toast)
        }

    }

    override fun onStart() {
        super.onStart()
        Log.i("QuizActivity", "onStart()")
    }

    override fun onResume() {
        super.onResume()
        Log.i("QuizActivity", "onResume()")
    }

    override fun onPause() {
        super.onPause()
        Log.i("QuizActivity", "onPause()")
    }

    override fun onStop() {
        super.onStop()
        Log.i("QuizActivity", "onStop()")
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.i("QuizActivity", "onDestroy()")
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        Log.i("QuizActivity", "onSaveInstanceState")
        outState?.putInt(KEY_INDEX, currentIndex)
    }
}

